### PR TITLE
test: Test source/sink status reporting

### PR DESCRIPTION
**Pull request by philip-stoev**
A mzcompose-based test that introduces disruptions against a source and a sink and confirms that the disruption is properly reflected in the mz_*_status tables. The disruption is then removed and the test checks that the status tables reflect resumed operation.

### Motivation

  * This PR adds a known-desirable feature.

Relates to #12864

### CODE REVIEW

File: test/source-sink-errors/mzcompose.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+                $ kafka-create-topic topic=source-topic
+
+                $ kafka-ingest topic=source-topic format=bytes
+                ABC
+
+                > CREATE SOURCE source1
+                  FROM KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-source-topic-${testdrive.seed}')
+                  FORMAT BYTES
+                  ENVELOPE NONE
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
```
Comment (philip-stoev) [inline test/source-sink-errors/mzcompose.py:73]: @bkirwi Using the `WITH` clause to connect to a storaged that has been started previously results in an empty status table.

File: test/source-sink-errors/mzcompose.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+                > CREATE SINK sink1 FROM source1
+                  INTO KAFKA CONNECTION kafka_conn (TOPIC 'testdrive-sink-topic-${testdrive.seed}')
+                  FORMAT AVRO USING CONFLUENT SCHEMA REGISTRY CONNECTION csr_conn
+                  ENVELOPE DEBEZIUM
+                  /* WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
+
+                # Wait for the source to start up
```
Comment (philip-stoev) [inline test/source-sink-errors/mzcompose.py:81]: @bkirwi If the disruption happens immediately after `CREATE SOURCE` succeeds, without waiting for the first data to arrive, the status table never transitions away from the `started` state, even though the logs contains the expected `librdkafka` errors.

It is true that in some sense the source never properly got "running", but being stuck in the "started" state and not in "stalled" means that the error column is empty even though librdkafka is already reporting the error in the log.

File: test/source-sink-errors/mzcompose.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        # Validate that cluster2 continues to operate
+        c.testdrive(
+            dedent(
+                f"""
+                > SELECT status, error ~* '{error}'
+                  FROM mz_internal.mz_source_status
+                  WHERE name = 'source1'
+                stalled true
+
+                #> SELECT status, error ~* '{error}'
```
Comment (philip-stoev) [inline test/source-sink-errors/mzcompose.py:98]: @bkirwi As discussed previously, sink errors do not appear to be reported.

File: test/source-sink-errors/mzcompose.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        c.testdrive(
+            dedent(
+                """
+                $ kafka-ingest topic=source-topic format=bytes
+                ABC
+
+                > SELECT COUNT(*) FROM source1;
+                2
+
+                #> SELECT status, error
```
Comment (philip-stoev) [inline test/source-sink-errors/mzcompose.py:116]: @bkirwi After the disruption has been cleared, the source is back up and ingesting , as evidenced by the `SELECT COUNT(*) = 2` check above. However, the status does not always transition to "running".

File: test/source-sink-errors/mzcompose.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+        )
+
+
+disruptions = [
+    Disruption(
+        name="delete-topic",
+        breakage=lambda c, seed: redpanda_topics(c, "delete", seed),
+        expected_error="UnknownTopicOrPartition",
+        fixage=None
+        # Re-creating the topic does not restart the source
```
Comment (philip-stoev) [inline test/source-sink-errors/mzcompose.py:136]: This is out of scope of this feature, but it seems that re-creating the topic does not cause ingestion to resume

Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:136]: That sounds right, though I'm glad we can at least test that the topic-deleted error percolates up.

File: test/source-sink-errors/mzcompose.py (hunk 4 above)
Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:116]: Investigating this; might be a timing issue.

File: test/source-sink-errors/mzcompose.py (hunk 3 above)
Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:98]: Working on this in https://github.com/MaterializeInc/materialize/pull/16232

File: test/source-sink-errors/mzcompose.py (hunk 2 above)
Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:81]: I agree that `stalled` would be better here... looking into it.

File: test/source-sink-errors/mzcompose.py (hunk 1 above)
Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:73]: Very puzzling; storaged should behave exactly the same in either case. Perhaps they're being launched with different configurations? I'll look into it.

File: test/source-sink-errors/mzcompose.py (hunk 4 above)
Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:116]: I think I've figured this one out: an event ordering issue in a multi-worker environment. I suspect most of my local testing was on a single worker.

I'll work up a fix to this, though I'm not sure exactly how to approach it yet.

File: test/source-sink-errors/mzcompose.py (hunk 1 above)
Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:73]:
```suggestion
                  /* TODO(bkirwi) WITH ( REMOTE 'storaged:2100' ) REMOTE causes the status source to be empty */
```

File: test/source-sink-errors/mzcompose.py (hunk 2 above)
Comment (bkirwi) [inline test/source-sink-errors/mzcompose.py:81]:
```suggestion
                # Wait for the source to start up
                # TODO(bkirwi) - catch errors while blocked on client startup
```

File: test/source-sink-errors/mzcompose.py (hunk 3 above)
Comment (bkirwi) [reply test/source-sink-errors/mzcompose.py:98]:
```suggestion
                # TODO(bkirwi) https://github.com/MaterializeInc/materialize/pull/16232
                #> SELECT status, error ~* '{error}'
```

File: test/source-sink-errors/mzcompose.py (hunk 4 above)
Comment (bkirwi) [inline test/source-sink-errors/mzcompose.py:116]:
```suggestion
                # TODO(bkirwi) https://github.com/MaterializeInc/materialize/issues/16065
                #> SELECT status, error
```

**Review by bkirwi — APPROVED** [review]
Thank you for this!

I've suggested a bunch of comments that assign future work to me. I propose merging this, and then in future bugfix PRs I can improve the test directly.

**Comment by philip-stoev** [comment]
I have incorporated your changes and squashed them in . Will merge the test on green. Please let me know if you need me on tracking any of the residual issues.

I would like to do another round of testing after all the issues have been moved out of the way. In particular, I would also like to do a short cloudtest/K8s test that checks that crash loops are detected properly (if in scope of the task).

**Pull request merged**